File: rht.js
import fs from "fs/promises";
import sqlite3 from "sqlite3";
import { open } from "sqlite";
import { initializeApp, cert } from "firebase-admin/app";
import { getFirestore, FieldValue, Timestamp } from "firebase-admin/firestore";

import serviceAccount from "./service_account-cp4.json" with { type: "json" };

initializeApp({
  credential: cert(serviceAccount),
});

const dbFirestore = getFirestore();

const DB_PATH = "/home/pi/Project/tastiway.db";
const COLLECTION_NAME = "tastiway_sensors";

async function readUnuploaded() {
  const db = await open({
    filename: DB_PATH,
    driver: sqlite3.Database,
  });

  const rows = await db.all(`
    SELECT id, doc, timestamp, rh, temp
    FROM rht
    WHERE upload = 0
    ORDER BY timestamp ASC
    LIMIT 144
  `);

  await db.close();
  return rows;
}

async function markUploadedMany(ids) {
  if (!ids || ids.length === 0) return;

  const db = await open({
    filename: DB_PATH,
    driver: sqlite3.Database,
  });

  const placeholders = ids.map(() => "?").join(",");

  await db.run(
    `UPDATE rht SET upload = 1 WHERE id IN (${placeholders})`,
    ids
  );

  await db.close();
}

function prepareEntries(rows) {
  const grouped = {};
  const rowIds = {};
  const timestamps = {};

  for (const row of rows) {
    const { id, doc, timestamp, rh, temp } = row;

    const ts = Timestamp.fromMillis(Number(timestamp));

    const entry = {
      timestamp: ts,
      rh,
      temp,
    };

    if (!grouped[doc]) {
      grouped[doc] = [];
      rowIds[doc] = [];
      timestamps[doc] = [];
    }

    grouped[doc].push(entry);
    rowIds[doc].push(id);
    timestamps[doc].push(ts);
  }

  return { grouped, rowIds, timestamps };
}

async function uploadGroupedToFirestore(grouped, timestamps) {
    const idPath = new URL("./rht_id.txt", import.meta.url);
    const ID = await fs.readFile(idPath, "utf8");

  for (const docId of Object.keys(grouped)) {
    const entries = grouped[docId];
    const latestTs = timestamps[docId].reduce((a, b) =>
      a.toMillis() > b.toMillis() ? a : b
    );

    const docRef = dbFirestore.collection(COLLECTION_NAME).doc(docId);

    try {
      await docRef.update({
        data: FieldValue.arrayUnion(...entries),
        timestamp: latestTs,
      });

      console.log(`Updated ${docId} with ${entries.length} entries`);
    } catch (err) {
      const msg = String(err);

      console.log(err);

      if (msg.includes("No document to update")) {
        // Create new document
        await docRef.set({
          data: entries,
          timestamp: latestTs,
          id: ID,
        });

        console.log(`Created new doc ${docId} with ${entries.length} entries`);
      } else {
        console.log(`Error updating ${docId}:`, err);
        continue;
      }
    }
  }
}

async function rht() {
  const rows = await readUnuploaded();

  if (!rows || rows.length === 0) {
    console.log("No pending records.");
    return;
  }

  const { grouped, rowIds, timestamps } = prepareEntries(rows);
  await uploadGroupedToFirestore(grouped, timestamps);

  const allIds = Object.values(rowIds).flat();
  await markUploadedMany(allIds);

  console.log("Done (batched).");
}

rht();

File: rht.py
import sqlite3
import firebase_admin
from firebase_admin import credentials, firestore
from datetime import datetime, timezone
from collections import defaultdict

cred = credentials.Certificate("./service_account-cp4.json")
firebase_admin.initialize_app(cred)
db = firestore.client()

DB_PATH = "/home/pi/Project/tastiway.db"
COLLECTION_NAME = "tastiway_sensors"

def read_unuploaded():
    conn = sqlite3.connect(DB_PATH)
    cur = conn.cursor()

    cur.execute("""
        SELECT id, doc, timestamp, rh, temp
        FROM rht
        WHERE upload = 0
        ORDER BY timestamp ASC
        LIMIT 144
    """)
    rows = cur.fetchall()

    conn.close()
    return rows

def mark_uploaded_many(ids):
    if not ids:
        return
    conn = sqlite3.connect(DB_PATH)
    cur = conn.cursor()
    cur.execute(
        f"UPDATE rht SET upload = 1 WHERE id IN ({','.join(['?']*len(ids))})",
        ids
    )
    conn.commit()
    conn.close()

def prepare_entries(rows):
    grouped = defaultdict(list)
    row_ids = defaultdict(list)
    timestamps = defaultdict(list)

    for row_id, doc, ts_ms, rh, temp in rows:

        ts = datetime.fromtimestamp(ts_ms / 1000, tz=timezone.utc)

        entry = {
            "timestamp": ts,
            "rh": rh,
            "temp": temp,
        }

        grouped[doc].append(entry)
        row_ids[doc].append(row_id)
        timestamps[doc].append(ts)

    return grouped, row_ids, timestamps


def upload_grouped_to_firestore(grouped, timestamps):
    with open("rht_id.txt", "r") as f:
        ID = f.read()
        print(ID)

    for doc_id, entries in grouped.items():
        # doc_ref = db.collection(COLLECTION_NAME).document(f'{doc_id}-{ID}')
        doc_ref = db.collection(COLLECTION_NAME).document(doc_id)

        # pick latest timestamp for top-level field
        latest_ts = max(timestamps[doc_id])

        try:
            doc_ref.update({
                "data": firestore.ArrayUnion(entries),
                "timestamp": latest_ts
            })
            print(f"Updated {doc_id} with {len(entries)} entries")

        except Exception as e:
            msg = str(e)
            print(e)
            if "404" in msg and "No document" in msg:
                doc_ref.set({
                    "data": entries,
                    "timestamp": latest_ts,
                    "id" : ID
                })
                print(f"Created new doc {doc_id} with {len(entries)} entries")
            else:
                print(f"Error updating {doc_id}: {e}")
                # do NOT set doc for any other unexpected error
                continue


def main():
    rows = read_unuploaded()

    if not rows:
        print("No pending records.")
        return

    grouped, row_ids, timestamps = prepare_entries(rows)
    upload_grouped_to_firestore(grouped, timestamps)

    # mark everything uploaded
    all_ids = [rid for sub in row_ids.values() for rid in sub]
    mark_uploaded_many(all_ids)

    print("Done (batched).")


if __name__ == "__main__":
    main()